Redundant colouring

rupertford opened the pull request:
Colouring and redundant computation transformations appear to work together now using the new colouring API. I've made PSyclone raise an exception when attempting to colour a loop that modifies a discontinuous field. Support for this can be added but I've had to ask a question to the LFRic developers on how the API works in this case.

Ready for first review by @arporter

arporter commented on src/psyclone/tests/test_files/dynamo0p3/9.1_orientation2.f90:
```diff
@@ -0,0 +1,48 @@
+! ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+! POSSIBILITY OF SUCH DAMAGE.
+! -----------------------------------------------------------------------------
+! Authors R. Ford and A. R. Porter, STFC Daresbury Lab
```
s/R. Ford/R. W. Ford/ and maybe drop me - I don't think I wrote this file originally?

rupertford replied:
Modified  my name and removed yours!

arporter commented on src/psyclone/tests/test_files/dynamo0p3/columnwise_op_asm_field_kernel_mod.F90:
```diff
@@ -1,18 +1,48 @@
-! Modifications copyright (c) 2017, Science and Technology Facilities Council
```
Corrections to make this kernel compilable? Great.

arporter commented on src/psyclone/tests/test_files/dynamo0p3/columnwise_op_asm_field_kernel_mod.F90:
```diff
@@ -100,22 +130,24 @@ subroutine columnwise_op_asm_field_kernel_code(cell,              &
+    real(kind=r_def), dimension(:,:,:), intent(in) :: local_stencil
     integer(kind=i_def), intent(in) :: nrow, ncol, bandwidth
-    real(kind=r_def), dimension(bandwidth,nrow,ncell_2d), intent(out) :: columnwise_matrix
+    real(kind=r_def), dimension(:,:,:), intent(out) :: columnwise_matrix
```
Did the compiler not like having explicit dimensions for the array? Ditto `local_stencil`?

rupertford replied:
I'm not sure. I cut it down from a Met Office example so either it was like that or I just didn't bother adding them. I've replaced the argument list with what we produce from kernelgen as the arguments they provided differ from the ones we generate!

arporter commented on src/psyclone/tests/test_files/dynamo0p3/infrastructure/mesh_mod.f90:
```diff
@@ -564,6 +599,16 @@ subroutine get_colours(self, ncolours, ncells_per_colour, colour_map)
+    class(mesh_type), intent(in), target      :: self
+    integer(i_def), pointer                   :: colour_map(:,:)
+
+    colour_map => self%cells_in_colour
```
For consistency with other stub code, this should be `=> null()`

arporter commented on src/psyclone/tests/test_files/dynamo0p3/infrastructure/operator_mod.f90:
```diff
@@ -62,14 +62,35 @@ module operator_mod
 
   type, extends(base_operator_type) ::  columnwise_operator_proxy_type
-     integer :: d
+     !type( function_space_type ), pointer :: fs_to
```
Please remove these commented-out lines.

rupertford replied:
Done.

arporter commented on src/psyclone/tests/test_files/dynamo0p3/testkern_orientation2.F90:
```diff
@@ -0,0 +1,56 @@
+! ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+! POSSIBILITY OF SUCH DAMAGE.
+! -----------------------------------------------------------------------------
+! Authors R. Ford and A. R. Porter, STFC Daresbury Lab
```
s/R. Ford/R. W. Ford/

rupertford replied:
Already done in a previous modification.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -353,6 +353,20 @@ def __init__(self, omp_schedule="static"):
         Transformation.__init__(self)
 
+    def _validate(self, node):
+        ''' Perform validation checks before applying the transformation'''
```
Please fully document the interface to this routine.

rupertford replied:
OK, done.

arporter replied:
Err, it's not! (You've done the one above but not the `_validate` at line 372.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -353,6 +353,20 @@ def __init__(self, omp_schedule="static"):
+                                      "must be computed serially.")
+
     def apply(self, node, reprod=None):
         '''Apply the OMPLoopTrans transformation to the specified node in a
```
Please update the docstring with the interface to this routine.

rupertford replied:
Done.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -379,22 +393,12 @@ def apply(self, node, reprod=None):
             import psyclone.config
             reprod = psyclone.config.REPRODUCIBLE_REDUCTIONS
 
-        # Check that the supplied node is a Loop
```
Moved to new `_validate()` routine - nice.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -1250,34 +1269,67 @@ def _validate(self, node, depth):
                 "In the Dynamo0p3RedundantComputation transformation apply "
-                "method the loop must iterate over cells or dofs, but found "
-                "'{0}'".format(node.loop_type))
+                "method the loop must iterate over cells, dofs or colour, "
```
again s/colour/cells of a single colour/ or some such?

rupertford replied:
already changed due to previous comment on this from the associated test

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -116,7 +116,8 @@
 
 VALID_LOOP_BOUNDS_NAMES = ["start", "inner", "cell_halo", "ncolour",
-                           "ncolours", "ncells", "ndofs", "dof_halo"]
+                           "ncolours", "ncells", "ndofs", "dof_halo",
```
I've added a list as suggested.

rupertford replied:
I've also updated the logic associated with loop bounds, particularly for colour_halo and added some new tests. PSyclone should now work with redundant computation and colouring.

rupertford commented:
When adding the new HALO_ACCESS_LOOP_BOUNDS list, as suggested by the reviewer, I went through all the places that loop bounds are used in logic within dynamo0p3.py. In some cases it is not clear to me that the correct thing will be done when there is redundant computation and so I am going to review the logic, make changes where appropriate and add new checks where required.

The particular issue I have found that causes concern is that I would expect the following line ...

`
        self._dirty_outer = (not field.discontinuous and
                             loop.upper_bound_name == "cell_halo")
`

(line 3385) to include a check for colour_halo as well.

Given that this is incorrect, why do the tests not fail? This bit of logic is used to change the bounds of what has been made clean (reducing it by 1). This functionality may be required when working out halo exchange bounds and when adding set_clean functions. However, the above logic is only used in the former and separate logic is used for set_clean. It turns out that the tests check for set_clean (which works) but not halo_exchange (which I believe will not).

I think it would be much better to use the same logic in both places so will look into that. However, in the first instance I will add a test that illustrates halo_exchange failing in particular cases where there is redundant computation.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -2709,15 +2710,19 @@ def gen_code(self, parent):
-            invoke_sub.add(DeclGen(parent, datatype="integer",
+            # Declare the colour map
+            declns = ["cmap(:,:)"]
+            if not config.DISTRIBUTED_MEMORY:
```
OK, done. Distributed memory calls a function for the upper bound so does not require a variable to be declared. This is the API provided by the Met Office. However, perhaps this API should be changed in the future to avoid dereferencing near parallel code (as this may cause problems with directives). I believe that this was the reason for the code being the way it is in the non-distributed memory version.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -2709,15 +2710,19 @@ def gen_code(self, parent):
-            invoke_sub.add(DeclGen(parent, datatype="integer",
-                                   entity_decls=["ncolour"]))
+                                   entity_decls=declns))
+            if not config.DISTRIBUTED_MEMORY:
```
I did not join them as it would have resulted in a different declaration ordering which would potentially affect a number of tests. Instead I explained why there is a difference between distributed memory and not, in the two places separately.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3732,9 +3738,22 @@ def _upper_bound_fortran(self):
 
         if self._upper_bound_name == "ncolours":
-            return "ncolour"
+            if config.DISTRIBUTED_MEMORY:
```
Done.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3732,9 +3738,22 @@ def _upper_bound_fortran(self):
+                return "ncolour"
         elif self._upper_bound_name == "ncolour":
             return "ncp_colour(colour)"
+        elif self._upper_bound_name == "colour_halo":
```
Done

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3732,9 +3738,22 @@ def _upper_bound_fortran(self):
+            append = ""
+            if halo_index:
+                append = ","+halo_index
+            return ("{0}%get_last_halo_cell_per_colour(colour"
```
Done.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -4057,13 +4080,16 @@ def gen_code(self, parent):
+                                              "ncolours"]:
                     for field in fields:
-                        if self._upper_bound_halo_depth:
+                        loop = field.call.parent
```
Yes I think it was a bug fix. However this piece of code has now been replaced.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -4488,12 +4515,21 @@ def gen_code(self, parent):
+            mesh_obj_name = self._name_space_manager.create_name(
+                root_name="mesh", context="PSyVars", label="mesh")
+            if config.DISTRIBUTED_MEMORY:
+                new_parent.add(AssignGen(new_parent, pointer=True, lhs="cmap",
```
Done.

rupertford commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -577,14 +629,14 @@ def test_check_seq_colours_omp_do():
     for dist_mem in [False, True]:
         psy = PSyFactory(TEST_API, distributed_memory=dist_mem).create(info)
-        invoke = psy.invokes.get('invoke_0_testkern_orientation_type')
+        invoke = psy.invokes.get('invoke_0_testkern_orientation2_type')
```
Done

rupertford commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -607,23 +660,20 @@ def test_colouring_after_openmp():
     _, info = parse(os.path.join(os.path.dirname(os.path.abspath(__file__)),
                                  "test_files", "dynamo0p3",
-                                 "9_orientation.f90"),
+                                 "1_single_invoke.f90"),
```
This is because 9_orientation calls the testkern_orientation kernel which writes to a discontinuous field. In this PR we have disabled support for colouring discontinuous loops so this test would no longer work. I could have changed to testkern_orientation2 (which is done in the previous loop) but testkern in 1_single_invoke is a simpler kernel so I chose that.

rupertford commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -1090,54 +1143,75 @@ def test_fuse_colour_loops():
-        assert call_idx2 > call_idx1
-        assert call_idx1 < end_loop_idx1
-        assert call_idx2 < end_loop_idx2
+        if dist_mem:
```
Done.

rupertford commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -3690,7 +3764,7 @@ def test_rc_invalid_loop(monkeypatch):  # pylint: disable=invalid-name
         rc_trans.apply(loop)
     assert ("In the Dynamo0p3RedundantComputation transformation apply "
-            "method the loop must iterate over cells or dofs, but "
+            "method the loop must iterate over cells, dofs or colour, but "
```
OK, done.

rupertford commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -3728,24 +3802,6 @@ def test_rc_invalid_depth():  # pylint: disable=invalid-name
             "the supplied depth is less than 1") in str(excinfo)
 
 
-def test_rc_invalid_depth_continuous():  # pylint: disable=invalid-name
```
As we now support redundant computation the error has changed to be that the depth must increase, rather than be greater than 1. This test has been replaced with test_rc_colour_no_loop_decrease().

rupertford commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -4450,33 +4506,6 @@ def test_rc_no_loop_decrease():  # pylint: disable=invalid-name
-    computation transformation after adding parallel directives (or in
-    general anything that becomes a parent of the loop). This is a
-    limitation that could be fixed and is down to the way we place new
-    halos (we put them before the loop and don't check whether there are
```
No, the restriction is still there. The way the errors are reported was changed and this test has been replaced by test_rc_parent_loop_colours_2()

rupertford commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -4886,3 +4915,526 @@ def test_red_comp_w_to_n_r_clean_gt_cleaned():  # pylint: disable=invalid-name
+def test_rc_parent_loop_colour(monkeypatch):
+    '''If the parent of the loop supplied to the redundant computation
+    transformation is a loop then the supplied loop should iterate
+    over 'colour'. If this is not the case then an exception is
```
Done.

rupertford commented:
OK, ready for second review by @arporter

arporter commented on src/psyclone/tests/test_files/dynamo0p3/infrastructure/operator_mod.f90:
```diff
@@ -62,14 +62,33 @@ module operator_mod
 
   type, extends(base_operator_type) ::  columnwise_operator_proxy_type
-     integer :: d
+     integer(i_def) :: alpha
```
I hadn't realised we had no compilation tests for CMA! Thanks for adding this stub infrastructure.

arporter commented on src/psyclone/tests/test_files/dynamo0p3/ru_kernel_mod.f90:
```diff
@@ -1,29 +1,102 @@
+      REAL(KIND=r_def), intent(in), dimension(1,ndf_w3,np_xy,np_z) :: basis_w3
+      INTEGER, intent(in), dimension(ndf_w0) :: map_w0
+      REAL(KIND=r_def), intent(in), dimension(1,ndf_w0,np_xy,np_z) :: basis_w0
+      REAL(KIND=r_def), intent(in), dimension(3,ndf_w0,np_xy,np_z) :: diff_basis_w0
```
Changes to make kernel compilable? Thanks.

rupertford replied:
Yes that's right.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -84,7 +84,23 @@ def name(self):
+        '''Perform various checks to ensure that it is valid to apply the
+        LoopFuseTrans transformation to the supplied nodes
+
+        :param node1: the first node we are checking
```
Great interface documentation - thanks!

arporter commented on src/psyclone/transformations.py:
```diff
@@ -1250,34 +1296,67 @@ def _validate(self, node, depth):
 
         if depth is None:
-            if node.upper_bound_name in ["cell_halo", "dof_halo"]:
+            if node.upper_bound_name in ["cell_halo", "dof_halo",
```
Please use `HALO_ACCESS_LOOP_BOUNDS` here. (Would `LOOP_BOUNDS_WITH_HALO_ACCESSES` be a better name?)

arporter commented on src/psyclone/transformations.py:
```diff
@@ -1304,14 +1383,8 @@ def _validate(self, node, depth):
-                    "1 as this loop modifies a continuous field")
-
-            if node.upper_bound_name in ["cell_halo", "dof_halo"]:
+            if node.upper_bound_name in ["cell_halo", "dof_halo",
```
Please use `HALO_ACCESS_LOOP_BOUNDS` here.

rupertford replied:
OK, done.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -1353,9 +1426,14 @@ def apply(self, loop, depth=None):
         if loop.loop_type == "":  # iteration space is cells
             loop.set_upper_bound("cell_halo", depth)
-        else:  # iteration space is dofs
+        elif loop.loop_type == "colour":
```
Please add comment, e.g. `# Loop is over cells of a single colour`

rupertford replied:
OK, done.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -115,8 +115,10 @@
 
-VALID_LOOP_BOUNDS_NAMES = ["start", "inner", "cell_halo", "ncolour",
-                           "ncolours", "ncells", "ndofs", "dof_halo"]
+HALO_ACCESS_LOOP_BOUNDS = ["cell_halo", "dof_halo", "colour_halo"]
```
Done.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -2911,7 +2922,7 @@ def _compute_halo_depth(self):
             if depth_info.max_depth:
                 # return the maximum halo depth
-                return "mesh%get_last_halo_depth()"
+                return "mesh%get_halo_depth()"
```
The API has changed. I've updated the comment to make it more clear.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -4035,10 +4082,11 @@ def gen_code(self, parent):
+                    # its parent loop
+                    hwa = HaloWriteAccess(field)
+                    if hwa.max_depth and not hwa.dirty_outer:
                         # do not output set dirty as it will all be
```
OK. I had it as it was as the logic seemed more understandable that way (to me). I've changed it to

`
if not hwa.max_depth or hwa.dirty_outer:
`

rupertford commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -4132,13 +4215,13 @@ def test_rc_dofs_no_depth():
         assert ("CALL {0}_proxy%halo_exchange(depth=mesh%"
-                "get_last_halo_depth())".format(field_name)) in result
+                "get_halo_depth())".format(field_name)) in result
     assert "DO df=1,f1_proxy%vspace%get_last_dof_halo()" in result
```
I've just checked and yes the api for dof halo depths is unchanged in the latest version of lfric trunk ( in gungho/source/function_space/get_last_dof_halo )

rupertford commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -4817,7 +4873,7 @@ def test_rc_max_w_to_r_continuous_known_halo():
 
     # sanity check that the halo exchange goes to the full halo depth
     assert ("w_to_r_halo_exchange._compute_halo_depth() == "
```
Well spotted!

rupertford commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -4885,3 +4941,1022 @@ def test_red_comp_w_to_n_r_clean_gt_cleaned():
+            "parent must iterate over 'colours'" in str(excinfo.value))
+
+
+def test_rc_parent_loop_colours_2(monkeypatch):
```
OK, if you insist. However, in general I am not sure when it is clearer to lay out each test and have replicated code, or when it is clearer to have multiple checks per test.

rupertford commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -4885,3 +4941,1022 @@ def test_red_comp_w_to_n_r_clean_gt_cleaned():
+
+    assert (
+        "      CALL f1_proxy%set_dirty()\n"
+        "      CALL f1_proxy%set_clean(1)" in result)
```
Added a comment similar to the one proposed - thanks.

rupertford commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -4885,3 +4941,1022 @@ def test_red_comp_w_to_n_r_clean_gt_cleaned():
+
+    # create our redundant computation transformation
+    rc_trans = Dynamo0p3RedundantComputationTrans()
+    # apply redundant computation to the colour loop
```
Done.

rupertford commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -4885,3 +4941,1022 @@ def test_red_comp_w_to_n_r_clean_gt_cleaned():
+    # create our redundant computation transformation
+    rc_trans = Dynamo0p3RedundantComputationTrans()
+
+    # apply redundant computation to the loop
```
Done.

rupertford commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -4885,3 +4941,1022 @@ def test_red_comp_w_to_n_r_clean_gt_cleaned():
+        "      DO colour=1,mesh%get_ncolours()\n"
+        "        DO cell=1,mesh%get_last_halo_cell_per_colour(colour,3)\n"
+        in result)
+
```
OK, done, although I think there are other tests that check for this (but not for this particular example)

rupertford commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -4885,3 +4941,1022 @@ def test_red_comp_w_to_n_r_clean_gt_cleaned():
+    # create our redundant computation transformation
+    rc_trans = Dynamo0p3RedundantComputationTrans()
+
+    # apply redundant computation to the loop
```
Done.

rupertford commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -4885,3 +4941,1022 @@ def test_red_comp_w_to_n_r_clean_gt_cleaned():
+
+    # pylint: disable=too-many-statements
+    def check_halo_exchange(halo_exchange):
+        '''internal method to check the validity of a particular halo
```
Done.

rupertford commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -4885,3 +4941,1022 @@ def test_red_comp_w_to_n_r_clean_gt_cleaned():
+        assert depth_info.literal_depth == 1
+        assert not depth_info.max_depth
+        assert not depth_info.var_depth
+
```
OK, done.

rupertford commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -4885,3 +4941,1022 @@ def test_red_comp_w_to_n_r_clean_gt_cleaned():
+        assert depth_info.max_depth
+        assert not depth_info.var_depth
+
+    _, invoke_info = parse(os.path.join(
```
Done.

rupertford commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -4885,3 +4941,1022 @@ def test_red_comp_w_to_n_r_clean_gt_cleaned():
+    # start with a loop which modifies the continuous field f1 to the
+    # maximum depth followed by a loop which modifies the continuous
+    # field f3 and reads field f1. This will produce a guaranteed halo
+    # exchange of maximum depth for field f1.
```
It was a cut and paste error which I've fixed. I've also changed to use a loop as requested.

rupertford commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -4885,3 +4941,1022 @@ def test_red_comp_w_to_n_r_clean_gt_cleaned():
+    schedule, _ = rc_trans.apply(w_loop, depth=2)
+    invoke.schedule = schedule
+    result = str(psy.gen)
+    assert result.count("f1_proxy%halo_exchange(") == 1
```
OK, done. I've also checked that the depth of the halo exchanges is 1 before redundant computation and 2 afterwards.

rupertford commented:
I think I've addressed all comments. Rather than restructuring of the tests into a new file I've created a new issue #130 which outlines some thoughts on what we could do. I've done this as restructuring the tests requires some thought.

rupertford commented:
Ready for 3rd review by @arporter.
